pl.from_epoch overflows on UInt32 input since 1.39.0

### Checks

- [x] I have checked that this issue has not already been reported.
- [x] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

epoch = 1721068200  # 2024-07-15 14:30:00 America/New_York

# UInt32
df_u32 = pl.DataFrame({"t": pl.Series([epoch], dtype=pl.UInt32)})
result_u32 = df_u32.select(pl.from_epoch("t", time_unit="s"))
print(f"UInt32: {result_u32['t'][0]}") 

# Int64
df_i64 = pl.DataFrame({"t": pl.Series([epoch], dtype=pl.Int64)})
result_i64 = df_i64.select(pl.from_epoch("t", time_unit="s"))
print(f"Int64:  {result_i64['t'][0]}") 

assert result_u32["t"][0] == result_i64["t"][0], (
    f"UInt32 and Int64 give different results: "
    f"{result_u32['t'][0]} != {result_i64['t'][0]}"
)
```
Output on Polars 1.39.3:
```
UInt32: 1970-01-01 00:29:50.048768
     Int64:  2024-07-15 18:30:00
AssertionError: UInt32 and Int64 give different results: 1970-01-01 00:29:50.048768 != 2024-07-15 18:30:00
```

### Log output

```shell

```

### Issue description

As of 1.39.3, `pl.from_epoch` performs seconds-to-microseconds multiplication in `UInt32` which overflows before widening to `UInt64`, so epoch values end up producing wrong results. The assertion in the provided test fails on 1.39.3.

### Expected behavior

The assertion passes: 
```
UInt32: 2024-07-15 18:30:00
     Int64:  2024-07-15 18:30:00
```

### Installed versions

1.39.3: 

```
--------Version info---------
Polars:              1.39.3
Index type:          UInt32
Platform:            macOS-26.2-arm64-arm-64bit
Python:              3.12.11 (main, Jun  3 2025, 15:41:47) [Clang 17.0.0 (clang-1700.0.13.3)]
Runtime:             rt32

----Optional dependencies----
Azure CLI            /opt/homebrew/Cellar/azure-cli/2.76.0/libexec/lib/python3.12/site-packages/azure/batch/models/_models.py:9067: SyntaxWarning: invalid escape sequence '\ '
  """The source port ranges to match for the rule. Valid values are '\ *' (for all ports 0 - 65535),
/opt/homebrew/Cellar/azure-cli/2.76.0/libexec/lib/python3.12/site-packages/azure/batch/models/_models.py:9235: SyntaxWarning: invalid escape sequence '\ '
  using brackets (for example abc[\ *] would match a file named abc*\ ). Note that both and / are
2.76.0
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                1.42.47
cloudpickle          3.1.2
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2026.2.0
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.10.8
numpy                2.4.2
openpyxl             <not installed>
pandas               3.0.0
polars_cloud         <not installed>
pyarrow              23.0.0
pydantic             2.12.5
pyiceberg            <not installed>
sqlalchemy           2.0.46
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

1.38.1:

```
UInt32: 2024-07-15 18:30:00
Int64:  2024-07-15 18:30:00
--------Version info---------
Polars:              1.38.1
Index type:          UInt32
Platform:            macOS-26.2-arm64-arm-64bit
Python:              3.12.11 (main, Jun  3 2025, 15:41:47) [Clang 17.0.0 (clang-1700.0.13.3)]
Runtime:             rt32

----Optional dependencies----
Azure CLI            2.76.0
adbc_driver_manager  <not installed>
altair               <not installed>
azure.identity       <not installed>
boto3                1.42.47
cloudpickle          3.1.2
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2026.2.0
gevent               <not installed>
google.auth          <not installed>
great_tables         <not installed>
matplotlib           3.10.8
numpy                2.4.2
openpyxl             <not installed>
pandas               3.0.0
polars_cloud         <not installed>
pyarrow              23.0.0
pydantic             2.12.5
pyiceberg            <not installed>
sqlalchemy           2.0.46
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

--- a/py-polars/src/polars/functions/lazy.py
+++ b/py-polars/src/polars/functions/lazy.py
@@ -2632,11 +2632,11 @@ def from_epoch(
     if time_unit == "d":
         return column.cast(Date)
     if time_unit in (scale := {"s": 1_000_000, "ms": 1_000}):
-        if isinstance(column, pl.Series) and column.dtype.is_integer():
-            column = column.cast(Int64)
-        elif isinstance(column, pl.Expr):
+        if isinstance(column, pl.Expr):
             column = column * F.lit(scale[time_unit], dtype=Int64)
             return column.cast(Datetime("us"))
+        if isinstance(column, pl.Series) and column.dtype.is_integer():
+            column = column.cast(Int64)
         return (column * scale[time_unit]).cast(Datetime("us"))
     if time_unit in DTYPE_TEMPORAL_UNITS:
         return column.cast(Datetime(time_unit))  # type: ignore[arg-type]

--- a/py-polars/src/polars/functions/lazy.py
+++ b/py-polars/src/polars/functions/lazy.py
@@ -2635,7 +2635,7 @@ def from_epoch(
         if isinstance(column, pl.Expr):
             column = column * F.lit(scale[time_unit], dtype=Int64)
             return column.cast(Datetime("us"))
-        if isinstance(column, pl.Series) and column.dtype.is_integer():
+        if column.dtype.is_integer():
             column = column.cast(Int64)
         return (column * scale[time_unit]).cast(Datetime("us"))
     if time_unit in DTYPE_TEMPORAL_UNITS:

--- a/py-polars/src/polars/functions/lazy.py
+++ b/py-polars/src/polars/functions/lazy.py
@@ -23,7 +23,7 @@ from polars._utils.parse import (
 from polars._utils.unstable import issue_unstable_warning, unstable
 from polars._utils.various import extend_bool, qualified_type_name
 from polars._utils.wrap import wrap_df, wrap_expr, wrap_s
-from polars.datatypes import DTYPE_TEMPORAL_UNITS, Date, Datetime
+from polars.datatypes import DTYPE_TEMPORAL_UNITS, Date, Datetime, Int64
 from polars.datatypes._parse import parse_into_datatype_expr
 from polars.lazyframe.opt_flags import (
     DEFAULT_QUERY_OPT_FLAGS,
@@ -2632,6 +2632,11 @@ def from_epoch(
     if time_unit == "d":
         return column.cast(Date)
     if time_unit in (scale := {"s": 1_000_000, "ms": 1_000}):
+        if isinstance(column, pl.Series) and column.dtype.is_integer():
+            column = column.cast(Int64)
+        elif isinstance(column, pl.Expr):
+            column = column * F.lit(scale[time_unit], dtype=Int64)
+            return column.cast(Datetime("us"))
         return (column * scale[time_unit]).cast(Datetime("us"))
     if time_unit in DTYPE_TEMPORAL_UNITS:
         return column.cast(Datetime(time_unit))  # type: ignore[arg-type]

--- a/py-polars/tests/unit/lazyframe/test_lazyframe.py
+++ b/py-polars/tests/unit/lazyframe/test_lazyframe.py
@@ -27,7 +27,11 @@ if TYPE_CHECKING:
 
     from _pytest.capture import CaptureFixture
 
-    from polars._typing import MapElementsStrategy, PolarsDataType
+    from polars._typing import (
+        EpochTimeUnit,
+        MapElementsStrategy,
+        PolarsDataType,
+    )
     from tests.conftest import PlMonkeyPatch
 
 
@@ -1346,14 +1350,65 @@ def test_from_epoch(input_dtype: PolarsDataType) -> None:
         _ = ldf.select(pl.from_epoch(ts_col, time_unit="s2"))  # type: ignore[call-overload]
 
 
-@pytest.mark.parametrize("input_dtype", [pl.UInt32, pl.Int32])
-def test_from_epoch_27107(input_dtype: PolarsDataType) -> None:
-    epoch_s = 1721068200  # 2024-07-15 18:30:00 UTC
-    ldf = pl.LazyFrame([pl.Series("timestamp_s", [epoch_s], dtype=input_dtype)])
-    res = ldf.select(pl.from_epoch("timestamp_s", time_unit="s"))
+@pytest.mark.parametrize(
+    ("input_dtype", "epoch_value", "time_unit", "expected_datetime"),
+    [
+        # 32-bit types with large positive values (original overflow case)
+        (pl.Int32, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        (pl.UInt32, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        # larger integer types
+        (pl.Int64, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        (pl.UInt64, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        (pl.Int128, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        (pl.UInt128, 1_721_068_200, "s", datetime(2024, 7, 15, 18, 30)),
+        # small unsigned types
+        (pl.UInt8, 100, "s", datetime(1970, 1, 1, 0, 1, 40)),
+        (pl.UInt16, 32_000, "s", datetime(1970, 1, 1, 8, 53, 20)),
+        # small signed types (positive values)
+        (pl.Int8, 100, "s", datetime(1970, 1, 1, 0, 1, 40)),
+        (pl.Int16, 32_000, "ms", datetime(1970, 1, 1, 0, 0, 32)),
+        # signed types with negative values (pre-epoch)
+        (pl.Int8, -100, "s", datetime(1969, 12, 31, 23, 58, 20)),
+        (pl.Int16, -32_000, "s", datetime(1969, 12, 31, 15, 6, 40)),
+        (pl.Int32, -1_721_068_200, "s", datetime(1915, 6, 19, 5, 30)),
+        (pl.Int64, -1_721_068_200, "s", datetime(1915, 6, 19, 5, 30)),
+        # milliseconds (with subsecond component)
+        (pl.Int64, 1_721_068_200_456, "ms", datetime(2024, 7, 15, 18, 30, 0, 456000)),
+        (pl.Int32, 2_000_456, "ms", datetime(1970, 1, 1, 0, 33, 20, 456000)),
+        (pl.Int64, -1_721_068_200_456, "ms", datetime(1915, 6, 19, 5, 29, 59, 544000)),
+        # nanoseconds (with subsecond component)
+        (
+            pl.UInt128,
+            1_721_068_200_456_789_000,
+            "ns",
+            datetime(2024, 7, 15, 18, 30, 0, 456789),
+        ),
+        (
+            pl.UInt128,
+            2_721_068_200_999_999_000,
+            "ns",
+            datetime(2056, 3, 23, 20, 16, 40, 999999),
+        ),
+        (
+            pl.Int128,
+            -1_721_068_200_456_789_000,
+            "ns",
+            datetime(1915, 6, 19, 5, 29, 59, 543211),
+        ),
+    ],
+)
+def test_from_epoch_27107(
+    input_dtype: PolarsDataType,
+    epoch_value: int,
+    time_unit: EpochTimeUnit,
+    expected_datetime: datetime,
+) -> None:
+    ldf = pl.LazyFrame({"ts": [epoch_value]}, schema={"ts": input_dtype})
+    res = ldf.select(pl.from_epoch("ts", time_unit=time_unit))
 
-    expected = pl.DataFrame([pl.Series("timestamp_s", [datetime(2024, 7, 15, 18, 30)])])
-    assert_frame_equal(res.collect(), expected)
+    dtype = pl.Datetime(time_unit if time_unit == "ns" else "us")
+    expected = pl.LazyFrame({"ts": [expected_datetime]}, schema={"ts": dtype})
+    assert_frame_equal(res, expected)
 
 
 def test_from_epoch_str() -> None:

--- a/py-polars/tests/unit/lazyframe/test_lazyframe.py
+++ b/py-polars/tests/unit/lazyframe/test_lazyframe.py
@@ -1346,6 +1346,16 @@ def test_from_epoch(input_dtype: PolarsDataType) -> None:
         _ = ldf.select(pl.from_epoch(ts_col, time_unit="s2"))  # type: ignore[call-overload]
 
 
+@pytest.mark.parametrize("input_dtype", [pl.UInt32, pl.Int32])
+def test_from_epoch_27107(input_dtype: PolarsDataType) -> None:
+    epoch_s = 1721068200  # 2024-07-15 18:30:00 UTC
+    ldf = pl.LazyFrame([pl.Series("timestamp_s", [epoch_s], dtype=input_dtype)])
+    res = ldf.select(pl.from_epoch("timestamp_s", time_unit="s"))
+
+    expected = pl.DataFrame([pl.Series("timestamp_s", [datetime(2024, 7, 15, 18, 30)])])
+    assert_frame_equal(res.collect(), expected)
+
+
 def test_from_epoch_str() -> None:
     ldf = pl.LazyFrame(
         [

